File: 6.mutex.py
import threading
import time

resource_mutex = threading.Lock()
readers_count_mutex = threading.Lock()
shared_resource = 0
readers = 0

def reader(id):
    global shared_resource, readers
    print(f"Reader {id} is trying to read")
    with readers_count_mutex:
        readers += 1
        if readers == 1:
            resource_mutex.acquire()  # Lock the resource for the first reader
    # Reading the shared resource
    print(f"Reader {id} read: {shared_resource}")
    time.sleep(1)
    with readers_count_mutex:
        readers -= 1
        if readers == 0:
            resource_mutex.release()  # Unlock the resource when the last reader is done

def writer(id):
    global shared_resource
    print(f"Writer {id} is trying to write")
    resource_mutex.acquire()
    # Writing to the shared resource
    shared_resource = id
    print(f"Writer {id} wrote: {shared_resource}")
    resource_mutex.release()

def main():
    num_readers = 4
    num_writers = 3

    reader_threads = []
    writer_threads = []

    for i in range(num_readers):
        reader_thread = threading.Thread(target=reader, args=(i+1,))
        reader_thread.start()
        reader_threads.append(reader_thread)

    for i in range(num_writers):
        writer_thread = threading.Thread(target=writer, args=(i+1,))
        writer_thread.start()
        writer_threads.append(writer_thread)

    for reader_thread in reader_threads:
        reader_thread.join()

    for writer_thread in writer_threads:
        writer_thread.join()

if __name__ == "__main__":
    main()

File: 6.mutex.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <unistd.h>

pthread_mutex_t resource_mutex;
pthread_mutex_t readers_count_mutex;
int shared_resource = 0;
int readers = 0;

void *reader(void *arg){
    int id = *((int *)arg);
    printf("Reader %d is trying to Read\n", id);
    pthread_mutex_lock(&readers_count_mutex);
    readers++;

    if (readers == 1){
        pthread_mutex_lock(&resource_mutex); // Lock the resource for the
        // first reader
    }

    pthread_mutex_unlock(&readers_count_mutex);
    // Reading the shared resource
    printf("Reader %d read: %d\n", id, shared_resource);
    usleep(1000000);
    pthread_mutex_lock(&readers_count_mutex);
    readers--;

    if (readers == 0){
        pthread_mutex_unlock(&resource_mutex); // Unlock the resource when the
        // last reader is done
    }
    pthread_mutex_unlock(&readers_count_mutex);
    pthread_exit(NULL);
}

void *writer(void *arg){
    int id = *((int *)arg);
    printf("Writer %d is trying to write\n", id);
    pthread_mutex_lock(&resource_mutex);

    // Writing to the shared resource
    shared_resource = id;
    printf("Writer %d wrote: %d\n", id, shared_resource);
    pthread_mutex_unlock(&resource_mutex);
    pthread_exit(NULL);
}
int main(){
    int num_readers = 4;
    int num_writers = 3;
    
    pthread_mutex_init(&resource_mutex, NULL);
    pthread_mutex_init(&readers_count_mutex, NULL);
    pthread_t reader_threads[num_readers];
    pthread_t writer_threads[num_writers];
    int reader_ids[num_readers];
    int writer_ids[num_writers];

    for (int i = 0; i < num_readers; i++){
        reader_ids[i] = i + 1;
        pthread_create(&reader_threads[i], NULL, reader, &reader_ids[i]);
    }

    for (int i = 0; i < num_writers; i++){
        writer_ids[i] = i + 1;
        pthread_create(&writer_threads[i], NULL, writer, &writer_ids[i]);
    }

    for (int i = 0; i < num_readers; i++){
        pthread_join(reader_threads[i], NULL);
    }

    for (int i = 0; i < num_writers; i++){
        pthread_join(writer_threads[i], NULL);
    }

    pthread_mutex_destroy(&resource_mutex);
    pthread_mutex_destroy(&readers_count_mutex);
    return 0;
}